tests/end_to_end_tests/container_image_tests/docker_test/conftest.py

```diff
@@ -0,0 +1,161 @@
+
+
+@pytest.fixture
+def dump_info(docker_server_hostname):
+    log.info(f"TEST INFO:\n" f"  server hostname: {docker_server_hostname}")
+
+
+def _call_docker(cmd_args: List[str]):
+    check_call_with_log(["docker", *cmd_args])
+
+
+@pytest.fixture(scope="session")
+def agent_container_name():
+    return "scalyr-agent"
+
+
+@pytest.fixture(scope="session")
+def docker_server_hostname(image_builder_name, test_session_suffix, request):
+    return f"agent-docker-image-test-{image_builder_name}-{request.node.nodeid}-{test_session_suffix}"
+
+
+@pytest.fixture
+def get_agent_log_content():
+    """
+    Fixture function that reads content of the agent log file in the agent's container.
+    """
+
+    def get(container_name: str) -> str:
+        return check_output_with_log(
+            [
+                "docker",
+                "exec",
+                "-i",
+                container_name,
+                "cat",
+                "/var/log/scalyr-agent-2/agent.log",
+            ],
+            description=f"Get content of the agent log in the container '{container_name}'.",
+        ).decode()
+
+    return get
+
+
+@pytest.fixture
+def start_agent_container(
+    image_name,
+    agent_container_name,
+    scalyr_api_key,
+    tmp_path_factory,
+    docker_server_hostname,
+    image_builder_name,
+    get_agent_log_content,
+):
+    """
+    Returns function which starts agent docker container.
+    """
+
+    # Kill and remove the previous container, if exists.
+    _call_docker(["rm", "-f", agent_container_name])
+
+    extra_config_path = tmp_path_factory.mktemp("extra-config") / "server_host.json"
+
+    extra_config_path.write_text(
+        json.dumps({"server_attributes": {"serverHost": docker_server_hostname}})
+    )
+
+    def start(timeout_tracker: TimeoutTracker):
+        # Run agent inside the container.
+
+        additional_options = []
+        # Add port if that's a syslog type image.
+        if "syslog" in image_builder_name:
+            additional_options.extend(["-p", "601:601"])
+
+        _call_docker(
+            [
+                "run",
+                "-d",
+                "--name",
+                agent_container_name,
+                "-e",
+                f"SCALYR_API_KEY={scalyr_api_key}",
+                "-v",
+                "/var/run/docker.sock:/var/scalyr/docker.sock",
+                "-v",
+                "/var/lib/docker/containers:/var/lib/docker/containers",
+                # mount extra config
+                "-v",
+                f"{extra_config_path}:/etc/scalyr-agent-2/agent.d/{extra_config_path.name}",
+                *additional_options,
+                image_name,
+            ]
+        )
+
+        with timeout_tracker(20):
+            while True:
+                try:
+                    get_agent_log_content(container_name=agent_container_name)
+                    return agent_container_name
+                except subprocess.CalledProcessError:
```

Comment:
Is there any way for us to provide additional context (e.g. in a somewhat generic manner) when timeout has been reached - things like logs accumulated so far, etc?

Reply:
Slightly changed this code. Now instead of checking the existence of the agent log, we use the `is_running` method of the existing class which controls the agent.

We now also accumulate logs in a background thread by using a new special fixture, which dumps that logs at the end of the test case (on failure too)